app/code/Magento/Backend/Controller/Adminhtml/Index/GlobalSearch.php

```diff
@@ -61,34 +75,64 @@ public function execute()
                         'Please make sure that all global admin search modules are installed and activated.'
                     ),
                 ];
-            } else {
-                $start = $this->getRequest()->getParam('start', 1);
-                $limit = $this->getRequest()->getParam('limit', 10);
-                $query = $this->getRequest()->getParam('query', '');
-                foreach ($this->_searchModules as $searchConfig) {
-                    if ($searchConfig['acl'] && !$this->_authorization->isAllowed($searchConfig['acl'])) {
-                        continue;
-                    }
-
-                    $className = $searchConfig['class'];
-                    if (empty($className)) {
-                        continue;
-                    }
-                    $searchInstance = $this->_objectManager->create($className);
-                    $results = $searchInstance->setStart(
-                        $start
-                    )->setLimit(
-                        $limit
-                    )->setQuery(
-                        $query
-                    )->load()->getResults();
-                    $items = array_merge_recursive($items, $results);
-                }
             }
         }
 
         /** @var \Magento\Framework\Controller\Result\Json $resultJson */
         $resultJson = $this->resultJsonFactory->create();
         return $resultJson->setData($items);
     }
+
+    /**
+     * @return array
+     */
+    protected function getPreviewItems()
+    {
+        $result = [];
+        $query = $this->getRequest()->getParam('query', '');
+        foreach ($this->previewModules as $previewConfig) {
+            if ($previewConfig['acl'] && !$this->_authorization->isAllowed($previewConfig['acl'])) {
+                continue;
+            }
+            if (!$previewConfig['url'] || !$previewConfig['text']) {
+                continue;
+            }
+            $result[] = [
+                'url' => $this->getUrl($previewConfig['url']).'?search='.$query,
+                'name' => __($previewConfig['text'], $query)
+            ];
+        }
+        return $result;
+    }
+
+    /**
+     * @return array
+     */
+    protected function getSearchItems()
+    {
+        $items = [];
+        $start = $this->getRequest()->getParam('start', 1);
+        $limit = $this->getRequest()->getParam('limit', 10);
+        $query = $this->getRequest()->getParam('query', '');
+        foreach ($this->_searchModules as $searchConfig) {
+            if ($searchConfig['acl'] && !$this->_authorization->isAllowed($searchConfig['acl'])) {
+                continue;
+            }
+
+            $className = $searchConfig['class'];
+            if (empty($className)) {
+                continue;
+            }
+            $searchInstance = $this->_objectManager->create($className);
```

Comment:
Marius, you right this looks like legacy code, even taking into account that it was written quite recently.

What we usually do for such cases is to introduce new Interface with the common contract, and Factory which is responsible for producing an instance of this interface and making sure that object which would be created is an instance of the interface mentioned above.
Factory also could serve as an extension point and other developers could introduce own instance of the contract and register it using DI configuration in this factory (constructor argument).

Doing so we get:

- well defined interface for some piece of functionality
- well defined Factory which creates some particular implementation of this interface
- extension point which helps 3rd party developers to extend basic implementations with custom one